File: answers/assignment03/pirallel.cpp
#include <iomanip>
#include <iostream>
#include <omp.h>

using namespace std;

double integrate(const int from, const int to, const double width) {
  double sum = 0.0;
  for (int i = from; i < to; i++) {
    const double x = (i + 0.5) * width;      // midpoint
    sum += (1.0 / (1.0 + x * x)); // add new height of a rectangle
  }
  return sum;
}

int main() {
  int num_steps = 100000000;              // amount of rectangles
  double width = 1.0 / double(num_steps); // width of a rectangle
  double start_time = omp_get_wtime(); // wall clock time in seconds

  double sum = integrate(0, num_steps, width); // for summing up all heights of rectangles

  double pi = sum * 4 * width; // compute pi
  double run_time = omp_get_wtime() - start_time;

  cout << "pi with " << num_steps << " steps is " << setprecision(17) << pi
       << " in " << setprecision(6) << run_time << " seconds\n";
}

File: answers/assignment01/pirallel.rs
use std::time::Instant;

fn main() {
    const NUM_STEPS: u32 = 100_000_000;
    const WIDTH: f64 = 1.0 / (NUM_STEPS as f64);

    let start_time = Instant::now(); // monotonically increasing clock

    let sum: f64 = (0..NUM_STEPS)
        .map(|i| ((i as f64) + 0.5) * WIDTH)
        .map(|x| 1.0 / (1.0 + x * x))
        .sum();

    let pi = sum * 4.0 * WIDTH;

    let run_time = (start_time.elapsed().as_millis() as f64) / 1000.0;

    println!(
        "pi with {:} steps is {:} in {:} seconds",
        NUM_STEPS, pi, run_time
    );
}
